Updates to connecting-to-a-portal (#9)
R-ArcGIS/r-bridge-site: location-services/connecting-to-a-portal.qmd

title: "Authorize with your Portal"
---

In order to create content or interact with non-public items in ArcGIS Online or ArcGIS Enterprise, you will need to be authorized using one of the following methods:

- [OAuth2](#authorization-with-oauth2)
- [Named User](#authorization-with-named-user)
- [ArcGIS Pro and `{arcgisbinding}`](#authorization-with-arcgis-pro-and-arcgisbinding)

:::{.callout-tip title="ArcGIS Sharing model" collapse="true"}
Items in ArcGIS Online and ArcGIS Enterprise can be unshared (accessible only to the item owner) or shared to groups, your organization, or everyone (![](images/sharing-public.png){width=30%}). [Learn more about Sharing here.](https://doc.arcgis.com/en/arcgis-online/share-maps/share-items.htm) Read and write access to an item is controlled by the owner-specified sharing level, organizational security settings, and the accessing user's [privileges](https://doc.arcgis.com/en/arcgis-online/administer/roles.htm).
:::

:::{.callout-warning}
If you modify environment variables or switch the active portal in ArcGIS Pro in the workflows below, you will need to restart your R session before proceeding.

> :::

Reviewer: Can we move this to the .Renviron section?

> :::{.callout-warning}
> If you modify environment variables or switch the active portal in ArcGIS Pro in the workflows below, you will need to restart your R session before proceeding.
> :::

Reviewer: ```suggestion
```

- **`ARCGIS_CLIENT`** and **`ARCGIS_SECRET`** environment variables are used only in the `auth_client()` OAuth2 client flow authorization. See the [article on authorization](https://r-arcgis.github.io/arcgislayers/articles/Authorization.html) for more. 

> - **`ARCGIS_TOKEN`** is used to fetch a user's access token. These are retrieved using the various `auth_` functions of `arcgisutils` and set using `set_arc_token()`. Alternatively, the `ARCGIS_TOKEN` can be set to an API key. 

Reviewer: ```suggestion
```
Let's delete this since it is no longer accurate after the token refactoring

:::

#### Setting your credentials in `.Renviron`

> Environment variables should _**never**_ be included in your code. Instead, they should be stored in an external environment file such as an `.Renviron` file. To modify your `.Renviron`, you can use the `usethis` package. Ensure it is installed and run the following in your text editor:

Reviewer: ```suggestion
Environment variables should _**never**_ be included in your code. Instead, they should be stored in an external environment file such as an `.Renviron` file. To modify your `.Renviron`, you can use the `usethis` package. Ensure it is installed and run the following in your text editor:

:::{.callout-warning}
If you modify environment variables you will need to restart your R session for the change to be registered.
:::
```

**User scoped** `.Renviron` files store environment variables in the user's home directory. The environment variables will be available to you in any project you open. However, if you share your project, the environment variables will not be shared with it.
:::

> This will open your `.Renviron` file for you to edit. Fill in the environment variables `ARCGIS_CLIENT` and `ARCGIS_SECRET` from the content item you just created. If you are not using ArcGIS Online, ensure that your portal's host is provided to `ARCGIS_HOST`. Additionally, provide your username to `ARCGIS_USER` if you plan to publish or edit existing content.

Reviewer: ```suggestion
This will open your `.Renviron` file for you to edit. Fill in the environment variables `ARCGIS_CLIENT` and `ARCGIS_SECRET` from the content item you just created. If you are not using ArcGIS Online, ensure that your portal's host is provided to `ARCGIS_HOST`. Additionally, provide your username to `ARCGIS_USER` if you plan to use the legacy username and password authentication.
```

This change because we are now inserting the username directly into the token field

```

:::{.callout-warning}
Authorization tokens are temporary and will expire. If you encounter an invalid token error, you might need to generate a new token.
:::

> We now need to set this token to the `ARCGIS_SECRET` environment variable. We do this with `set_arc_token()`. 

Reviewer: ```suggestion
To make this token easily accessible to `{arcgis}`, use `set_arc_token()` which sets the token in dedicated authorization token environment.
```

```r
set_arc_token(token)

> #> Token set to environment variable `ARCGIS_TOKEN`

Reviewer: ```suggestion
```

the message is not printed anymore unless we're doing the named token process

```

#### Client flow

> Alternatively, we can authorize using the client OAuth2 flow. This will authorize the application we created and not ourselves. Because of this, we cannot use the client flow to create or modify content. 

Reviewer: ```suggestion
Alternatively, you can authorize using the client OAuth2 flow. This will authorize the application you created and not ourselves. Because of this, you cannot use the client flow to create or modify content. 
```

The client flow has the benefit of being non-interactive, though. 

```r
token <- auth_client()
set_arc_token(token)

> #> Token set to environment variable `ARCGIS_TOKEN`

Reviewer: ```suggestion
```

```

## Authorization with Named User

> Authorizing using an ArcGIS Online or ArcGIS Enterprise [username and password](https://developers.arcgis.com/documentation/mapping-apis-and-services/security/user-authentication/) is a legacy method that is supported for cases where OAuth2 is not implemented. As a reminder, credentials should never be stored in plaintext in code.

Reviewer: ```suggestion
Authorizing using an ArcGIS Online or ArcGIS Enterprise [username and password](https://developers.arcgis.com/documentation/mapping-apis-and-services/security/user-authentication/direct-username-password/) is a legacy method that is supported for cases where OAuth2 is not implemented. As a reminder, credentials should never be stored in plaintext in code.

:::{.callout-important}
**Security consideration**: Obtaining an access token with this method will expose the username and password credentials as plain text and could present a potential security risk.
:::
```

```r
token <- auth_user(
  username = Sys.getenv("ARCGIS_USER"),
  password = Sys.getenv("ARCGIS_PASSWORD")
  host = arc_host(),
  expiration = 60
)
set_arc_token(token)

> #> Token set to environment variable `ARCGIS_TOKEN`

Reviewer: ```suggestion
```

```

## Authorization with ArcGIS Pro and `{arcgisbinding}`

If you are a user of ArcGIS Pro and have **arcgisbinding** installed, you can use `auth_binding()` to utilize the tokens that result from `arc.check_portal()`. `auth_binding()` has the benefit of being non-interactive _and_ authorizes you as a user. You can use `auth_binding()` for non-interactive work that creates or modifies existing content.

```r
token <- auth_binding()
set_arc_token(token)

> #> Token set to environment variable `ARCGIS_TOKEN`

Reviewer: ```suggestion
```
